src/quacc/recipes/espresso/phonons.py

```diff
@@ -0,0 +1,142 @@
+        scratch of this calculation. For ph_job this is what most users will want to do.
+
+        If a list of strings is provided, each string point to a specific file. In this case
+        it is important to note that no directory structure is going to be copied, everything
+        is copied at the root of the scratch directory.
+
+    parallel_info
+        Dictionary containing information about the parallelization of the
+        calculation. See the ASE documentation for more information.
+
+    **calc_kwargs
+        calc_kwargs dictionary possibly containing the following keys:
+
+        - input_data: dict
+        - qpts: list[list[float]] | list[tuple[float]] | list[float]
+        - nat_todo: list[int]
+
+        See the docstring of quacc.calculators.espresso.io.write_espresso_ph for more
+        information.
+
+    Returns
+    -------
+    RunSchema
+        Dictionary of results from [quacc.schemas.ase.summarize_run][]
+    """
+
+    calc_defaults = {
+        "input_data": {
+            "inputph": {
+                "tr2_ph": 1e-12,
+                "alpha_mix(1)": 0.1,
+                "nmix_ph": 12,
+                "verbosity": "high",
+            }
+        },
+        "qpts": (0, 0, 0),
+    }
+
+    template = EspressoTemplate("ph", test_run=test_run)
+
+    return base_fn(
+        Atoms(),
+        preset=preset,
+        template=template,
+        calc_defaults=calc_defaults,
+        calc_swaps=calc_kwargs,
+        parallel_info=parallel_info,
+        additional_fields={"name": "ph.x static"},
+        copy_files=copy_files,
+    )
+
+
+@flow
+def grid_phonon(pw_job: Job = static_job, phonon_job: Job = phonon_job):
+
+    # We run a first pw job (single point or relax) depending on the input
+    pw_job_results = pw_job()
+
+    # First ph job to generate the patterns
+    ph_job_partial = partial(
+        phonon_job, test_run=True, copy_files=pw_job_results["dir_name"]
+    )
+    # We acutally run the job
+    ph_job_results = ph_job_partial()
+    # We parse the patterns
+    ph_patterns = parse_ph_patterns(ph_job_results["dir_name"])
+    # We need to get the input_data from the ph job that was sent in...
+    input_data = ph_job_results["parameters"]["input_data"]  # Better way to do that?
+
+    grid_results = []
+    # We loop over the q-points and representations and run a ph job for each
+    # of them
+    for pattern in ph_patterns:
+        input_data["inputph"]["start_q"] = pattern
+        input_data["inputph"]["last_q"] = pattern
+        n_repr = ph_patterns[pattern]
+        for representation in range(1, n_repr + 1):
+            input_data["inputph"]["start_irr"] = representation
+            input_data["inputph"]["last_irr"] = representation
+            ph_job_partial = partial(
+                phonon_job, input_data=input_data, copy_files=pw_job_results["dir_name"]
+            )
+            grid_results.append(ph_job_partial())
```

Comment:
So, this may technically work in Covalent but is not a general approach. The reason for this is that `ph_patterns` is effectively a `Future` with unknown length until runtime, and so it is impossible to know _a priori_ how many iterations to loop over. This kind of pattern is described by the `@subflow` decorator in quacc, which meant when you are iterating over a bunch of `Job`s that you want to then collect.

Subflows (called sublattices in Covalent, join apps in Parsl, and subworkflows elsewhere) are generally the most difficult part of workflow management, so I'll take care of refactoring this appropriately when you're ready. You can see an example [here](https://github.com/Quantum-Accelerators/quacc/blob/main/src/quacc/recipes/common/slabs.py#L18).